crates/asm/subprotocols/checkpointing-v0/src/verification.rs

```diff
@@ -0,0 +1,156 @@
+        let expected = state.expected_next_epoch();
+        logging::warn!(expected, actual = epoch, "Invalid epoch progression");
+        return Err(CheckpointV0Error::InvalidEpoch {
+            expected,
+            actual: epoch,
+        });
+    }
+
+    ensure_batch_epochs_consistent(checkpoint)?;
+    verify_checkpoint_signature(signed_checkpoint, &state.cred_rule)?;
+    verify_checkpoint_proof(checkpoint, state)?;
+
+    if let Some(previous) = &state.last_checkpoint {
+        verify_state_transition(previous, checkpoint)?;
+    }
+
+    state.update_with_checkpoint(checkpoint.clone(), current_l1_height);
+    logging::info!(epoch, "Successfully verified checkpoint");
+
+    Ok(())
+}
+
+fn ensure_batch_epochs_consistent(checkpoint: &Checkpoint) -> Result<(), CheckpointV0Error> {
+    let info_epoch = checkpoint.batch_info().epoch();
+    let transition_epoch = checkpoint.batch_transition().epoch;
+    if info_epoch != transition_epoch {
+        return Err(CheckpointV0Error::StateTransitionError(
+            "batch info and transition epochs differ".to_string(),
+        ));
+    }
+    Ok(())
+}
+
+fn verify_checkpoint_signature(
+    signed_checkpoint: &SignedCheckpoint,
+    cred_rule: &CredRule,
+) -> Result<(), CheckpointV0Error> {
+    if verify_signed_checkpoint_sig(signed_checkpoint, cred_rule) {
+        Ok(())
+    } else {
+        Err(CheckpointV0Error::InvalidSignature)
+    }
+}
+
+fn verify_checkpoint_proof(
+    checkpoint: &Checkpoint,
+    state: &CheckpointV0VerifierState,
+) -> Result<(), CheckpointV0Error> {
+    let proof_receipt = checkpoint.construct_receipt();
+    let expected_output = *checkpoint.batch_transition();
+    let actual_output: BatchTransition =
+        borsh::from_slice(proof_receipt.public_values().as_bytes())
+            .map_err(|_| CheckpointV0Error::SerializationError)?;
+
+    if expected_output != actual_output {
+        logging::warn!(
+            epoch = checkpoint.batch_info().epoch(),
+            "Checkpoint proof public values mismatch"
+        );
+        return Err(CheckpointV0Error::InvalidProof);
+    }
+
+    let is_empty_proof = proof_receipt.proof().is_empty();
+    let allow_empty = state.proof_publish_mode.allow_empty()
+        || matches!(
+            state.rollup_verifying_key,
+            RollupVerifyingKey::NativeVerifyingKey
+        );
+
+    if is_empty_proof {
+        if allow_empty {
+            logging::warn!(
+                epoch = checkpoint.batch_info().epoch(),
+                "Accepting empty checkpoint proof"
+            );
+            return Ok(());
+        }
+
+        return Err(CheckpointV0Error::InvalidProof);
+    }
+
+    if let Err(err) =
+        verify_rollup_groth16_proof_receipt(&proof_receipt, &state.rollup_verifying_key)
+    {
+        logging::warn!("Groth16 verification failed: {err:?}");
+        return Err(CheckpointV0Error::InvalidProof);
+    }
+
+    Ok(())
+}
+
+fn verify_state_transition(
```

Comment:
There is actually no state transition here, right? We just check that the previous epoch (as declared in the checkpoint) ends where this epoch starts. Maybe something like _verify_continous_transition_? or _states_?